### ISSUE TITLE
When a user clicks the switch an event or method should be able to be called.

### ISSUE THREAD

**Issue by markalanevans** [frapontillo/angular-bootstrap-switch#80]
# Question
How can we bind an event to a toggle?

# Story
As a developer I should be able to trigger a method call when a switch is clicked.

**Comment by frapontillo** [comment]
...or you could set up a watcher on your model :)
BTW, this is not the first time this request has popped up, I'm leaving this thread open for a while to see what kind of support this one has.

**Comment by markalanevans** [comment]
sure. basically you just need to pass through the ngClick functionality.

**Comment by RatkoR** [comment]
Yup, event like this would be a big bonus. At least for me.

**Comment by BrOrlandi** [comment]
I've made an implementation of that functionality.
You can see in my commit: https://github.com/BrOrlandi/angular-bootstrap-switch/commit/d50589021c22295f267763240d789f9b33420683

The Angular watch doesn't work well in Angular 1.2, this implementation makes it more suitable with Angular 1.2.

Some tests are not passing. I don't have checked it, made this implementation just to show how it can work with Angular 1.2.

**Comment by konczak** [comment]
Hi, I would like to see it in place.
I'm struggling with dynamic list of objects where every one of them can toggle.
It is mandatory on click to switch state also in backend (I need to call REST API to do that).
Callback function would be a good solution for my case as long as I will be able to identify which object should change.

**Comment by BrOrlandi** [comment]
Hi @konczak my case is exactly the same as yours. The switch callback I've made receives the DOM element of the switch, so I've set an custom attribute with the id of the element I want to update via REST API. On the callback function I get this element attribute and do the REST request.

**Comment by frapontillo** [comment]
Could you please explain why you can't set up a watcher on the model bound to the switch? It's something I may have already asked before, but since we're trying to understand the issue here...
Thanks!

**Comment by RatkoR** [comment]
In my case, model's binded property can change because user flipped 
switch to on/off and it can also change in controller because of some 
other events.

If I add a watch, It'll catch both changes: switch changes and changes 
done in javascript. Which is not ok, since I need to process only change 
done by switch.

Thanks,
Ratko

Dne 07. 07. 2015 ob 08:02 je Francesco Pontillo zapisal(a):

**Comment by BrOrlandi** [comment]
Hi @frapontillo I can't use watchers because I'm using Angular 1.2 and they don't work very well at this version. On Angular 1.3 they work OK.
Also, a callback should be better than have a watch inside a loop through a list.

**Comment by konczak** [comment]
Hi @frapontillo for me is the same case as @BrOrlandi that temporary I'm stuck with Angular 1.2 without possibility to upgrade.

Also I prefer solution with callbacks then watches because personally I think that code gets messy with watches while with callback it is much more clear when it will be executed - callback gives better control over execution.

How about performance case? I think it is good enough when we speak about single switch. While in my case I need it for list which size is dynamic - watching for any change on it can be a bad impact on user experience and memory consumption.

**Comment by frapontillo** [comment]
What kind of arguments do you want your callback to have? The model the switch is bound to?

**Comment by konczak** [comment]
That's very good question. I'm worried that only model is insufficient. I need identify which object has changed state.
In my case I have list of Features where every single Feature has ID, name, enabled and other fields.
Field enabled is my flag used in switch as model.
I need at least ID which I will use while sending request to backend.
Probably most useful would be to have hole Feature object - ex. then I could easily (without searching array by ID) take name of the Feature and display popup that "Feature NAME is enabled now".

**Comment by BrOrlandi** [comment]
@frapontillo you can see my implementation https://github.com/BrOrlandi/angular-bootstrap-switch/commit/d50589021c22295f267763240d789f9b33420683
The Bootstrap Switch in jQuery has a callback, I've just used the same arguments.
They are: the DOM element, the event and the new state of the switch.
As I explained to @konczak I've used an custom attribute in the element to get the ID of the object that has changed the state.

**Comment by frapontillo** [comment]
@BrOrlandi I am not going to expose the DOM to the controller, that's something non-Angular-ly.

What I can do, though, is eval some specific code in a `switch-click` event, such as:

```html
<input type="checkbox"
    switch-click="hasBeenClicked(myElement)"
/>
```

This way, `hasBeenClicked` will be any method you wish and `myElement` any object in the current `$scope` (it can be an element iterated with `ng-repeat`, for example).

**Comment by BrOrlandi** [comment]
I'm not a expert in Angular, so I don't know the best practices.
The solution you've proposed it's good and should be enough.
But I suggest the nmae `switch-change` than `switch-click` to be suitable to the jQuery plugin.
Thanks.

**Comment by konczak** [comment]
I really like @frapontillo solution with customizable function and list of parameters - that will fix my issues greatly.
@BrOrlandi if it has to be switch-change let it be - it has good mapping with jQuery. For Angular compliance switch-click probably would be better but we can't have both.

I have one more question - when this function will be called? After change ng-model value?
I'm considering it like that: when user presses on switch it should call backend, if call completed successfully value of switch should change. If backend request failed from any reason it would be good to avoid change value of switch ng-model.
I know that I can achieve this easily setting ng-model value to old one - it will do the job in my case. Still user will experience switching animation twice.

**Comment by frapontillo** [comment]
I think it should really be a callback and nothing else, that kind of stuff should be handled with validation modules and techniques.

**Comment by SeanEire** [comment]
I'm looking for this exact feature too! @frapontillo your customizable function solution would be ideal.

**Comment by zaewonyx** [comment]
Hello @frapontillo, I'm interested too by this callback function support. Could you give a date for a possible release ?

**Comment by frapontillo** [comment]
@zaewonyx I honestly don't know when I'm going to have some time to allocate for this... Probably mid-to-end October. Yes, I really wrote October, unfortunately :disappointed:.

I would gladly accept well-written and well-tested PRs, though.

**Comment by zaewonyx** [comment]
Thank you for your fast response. I'll handle that with an ng-change for now and maybe I'll do a PR (I'm not sure of the well-written and well-tested part :blush:)

**Comment by ri0t** [comment]
I would love that functionality, too. I could possibly try with watchers, but my current use case involves ng-repeat with 150+ switches, so i'm a bit worried about performance impacts.. ..the app needs to conserve battery as much as possible.

**Comment by MtDalPizzol** [comment]
Well, I think that what the great majority of the developers need to do, while using this directive, is to get the scope's object bound to the element.

Also, I don't think setting up watchers will do the trick, because in some cases, you need to use the directive inside a ng-repeat...

So, the solution for me was this:

**angular-bootstrap-switch.js**
```javascript
// line 19
function($parse, $timeout) {
    return {
        restrict: 'A',
        require: 'ngModel',
        scope: {
          'switchChange': '&'
        },

// line 230
element.on('switchChange.bootstrapSwitch', function (e, state) {
    // $setViewValue --> $viewValue --> $parsers --> $modelValue
    controller.$setViewValue(e.target.checked);
    scope.switchChange();
});

// line 248
]).directive('bsSwitch', function() {
    return {
        restrict: 'E',
        require: 'ngModel',
        template: '<input bs-switch ng-click="switchChange({ item: item })">',
        replace: true
    };
});
```

**controller.js**
```javascript
app.controller('MyController', ['$scope', function($scope){
    $scope.updateState = function(item) {
        console.log(item);
    };
}]);
```
**index.html**
```
<input
    type="checkbox"
    ng-model="item.men_ativo"
    bs-switch
    switch-size="mini"
    switch-on-text="{{i18n.active}}"
    switch-off-text="{{i18n.inactive}}"
    switch-on-color="success"
    switch-off-color="danger"
    switch-change="updateState(item)">
```

I hope this help someone...

**Comment by konczak** [comment]
Is there any chance that will be a part of this awesome directive?

**Comment by jakerobers** [comment]
Is there an ETA for this? I noticed @frapontillo mentioned something about end of October? I would also like this feature.

**Comment by tanzaho** [comment]
Questions to those who previously posted: in the meantime, what is the solution are you using ? I'm also using a ng-repeat with many elements in it.

**Comment by adrienbrunet** [comment]
:+1:

**Comment by MtDalPizzol** [comment]
Hi @tanzaho , I'm still using the solution I posted above. I know it could be better, but it does the trick in a simple way and the team here at my company haven't ran into problems with that...

**Comment by icholy** [comment]
I think the existence of the built in `ngChange` directive suggests that this is valid use case.

**Comment by ArthurPBressan** [comment]
@frapontillo, please consider merging @icholy's  change. Using watchers to watch over changes isn't reasonable, mostly because of performance reasons.

**Comment by ArthurPBressan** [comment]
:+1: for @MtDalPizzol's solution.

**Comment by MtDalPizzol** [comment]
Hey everyone... Seems like a bunch of people are getting around the problem with the suggestion I made. I don't know if someone has made a PR or something, but if @frapontillo thinks it is valid, I can make one... What do you think?

**Comment by frapontillo** [comment]
Everyone: sorry for not addressing this issue for a long time, it's been kind of a busy time.

@MtDalPizzol there is already PR #97 that should address this, could you (and everyone else) please verify it matches your expectations?

Thank you for your patience.

**Comment by frapontillo** [comment]
#97 is now live in develop, could you try it out and see if it works as expected?

**Comment by tanzaho** [comment]
@frapontillo Thanks for the work on that issue. However, the script doesn't seem to fire as @ArthurPBressan said in #97.

**Comment by ArthurPBressan** [comment]
I tried to get a plunker working, but I couldn't get it to work and I'm not sure what I'm missing. [Here it is, anyway](https://plnkr.co/edit/LUNvxWGjbpg3tFnMC80N?p=preview). This plunker is using icholy's code.

@MtDalPizzol, can you get a fork up with your changes, please? It's probably going to make the eventual plunker easier to set up.

**Comment by icholy** [comment]
#99 fixes this issue. Here's a plunker with the latest version https://plnkr.co/edit/YUpJ8X2ZXN6XRTPkI7Ay?p=preview

**Comment by frapontillo** [comment]
@tanzaho @ArthurPBressan could you please check if #99 fixes it for good?

**Comment by ArthurPBressan** [comment]
Worked for me. :+1: 

Thanks a lot!

**Comment by tanzaho** [comment]
@frapontillo @icholy @ArthurPBressan By using the script provided in the Plunkr, it seems to work as expect on my project.

I can also confirm that switch-change can take parameters.

**Comment by frapontillo** [comment]
4c5fcc1 should now have the requested feature. Once again, could you please check if it is properly working?
Cheers @icholy for implementing this.

**Comment by tanzaho** [comment]
@frapontillo : I confirm that it works on my local project. :+1:

**Comment by ArthurPBressan** [comment]
Is the dist folder updated with this change?

**Comment by frapontillo** [comment]
Not just yet, as soon as I get the greenlight from the most of you I'll cut a release.

**Comment by icholy** [comment]
ping ^ @markalanevans @RatkoR @BrOrlandi @konczak @SeanEire @zaewonyx @ri0t @jakerobers

**Comment by icholy** [comment]
@frapontillo no news if good news? Please update dist files.

**Comment by konczak** [comment]
Hi, 
I confirm that on my local project this solution works for me as expected.

PS. Sorry for late answer.

**Comment by zaewonyx** [comment]
Hi, it works as expected for me too.

Thanks ! And sorry for not replying

**Comment by jjmontesl** [comment]
Found this thread while facing the same issue. Downloaded the development branch and it's working fine in my case. The name `switch-change` may be confusing perhaps, since it does not trigger on model change (as expected). I'd favour `switch-click` or `switch-interaction`. Thanks for this!

**Comment by frapontillo** [comment]
Released in 0.5.0 (dffb8ba9e7d2da1ef7c8d8b5b76063923b983188).
